tabular/src/autogluon/tabular/models/abstract/abstract_model.py

```diff
@@ -189,10 +189,21 @@ def rename(self, name: str):
         self.path = self.path[:-len(self.name) - 1] + name + os.path.sep
         self.name = name
 
-    # Extensions of preprocess must act identical in bagged situations, otherwise test-time predictions will be incorrect
-    # This means preprocess cannot be used for normalization
-    # TODO: Add preprocess_stateful() to enable stateful preprocessing for models such as KNN
-    def preprocess(self, X):
+    def preprocess(self, X, preprocess=True, preprocess_stateful=True, **kwargs):
+        if preprocess:
+            X = self._preprocess(X, **kwargs)
+        if preprocess_stateful:
+            X = self._preprocess_stateful(X, **kwargs)
+        return X
+
+    # TODO: Remove kwargs?
+    def _preprocess(self, X, **kwargs):
+        """
+        Data transformation logic that is non-stateful or ignores internal data values beyond feature dtypes should be added here.
```

Comment:
@jwmueller Curious on your thoughts on this. I'm open to changing to _preprocess_stateless (and argument to preprocess_stateless), but I also want to avoid things sounding overly complicated. Also, it's not technically 'stateless', it's just that anything that happens to be stateful in this method must not be stateful relative to the training data inner values.

Reply:
I agree stateless sounds misleading. I also agree the `preprocess` argname for `preprocess()` method would be clearer if changed to eg. `preprocess_fixed` or `preprocess_nonadaptive` or `preprocess_once_per_run` (stateless is too misleading IMO).

I'm more ambivalent whether the name of `_preprocess()` function is changed. The way it is right now could lead to some new model contributors overlooking `_process_stateful()`, but on the flip side, we do want most contributors to pack most of their preprocessing code into `_preprocess()` whenever possible (which they'll probably do under the current name) and to treat `_preprocess_stateful()` as a bonus method only extended when necessary.

Reply:
Thanks for the comment. I've gone and renamed _preprocess -> _preprocess_nonadaptive and _preprocess_stateful -> _preprocess, as well as argument name preprocess -> preprocess_nonadaptive (keeping preprocess_stateful argument).

With this, I want to position the stateful preprocessor as the go-to for most users, and nonadaptive as the advanced option we can add after initial implementation for a speedup.

Reply:
This means I've also moved RF and KNN to be stateful preprocessing, since they convert to numpy in their preprocessor which actually was breaking the inner feature importance model method (not usually used by users). I will benchmark to see if this is a significant inference slowdown. If it is, I can change it back for RF and KNN to be nonadaptive and remove the assumption that _preprocess_nonadaptive returns a DataFrame.